### DISCUSSION TITLE
How to Get the Current User with dependencies added to APIRouter

### DISCUSSION THREAD

**Discussion by Zaeem6100** [fastapi/fastapi#11150] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
#here is my router.py file 

from fastapi import APIRouter, Depends

from src.auth.dependencies import verify_user

router = APIRouter(
    prefix="/auth",
    tags=["auth"],
    dependencies=[Depends(verify_user)],
    responses={404: {"description": "Not found"}},
)


@router.get("/", tags=["auth"])
async def root():
    return {"message": "Hello auth"}



# dependency.py
async def verify_user(x_token: Annotated[str, Header()]) -> User:
    try:
        if x_token == "uza":
            #     todo: set the default user to a fake user
            return User(first_name="fake", username="faker", email="example@example.com", age=21)
        else:
            payload = jwt.decode(x_token, JWT_SECRET_KEY, algorithms=["HS256"])
            print(payload)
            return User(**payload)
    except JWTError as e:
        raise HTTPException(status_code=401, detail="Invalid token provided")
```

### Description

I have added the dependencies in the APIRouter. I want to auth the token at the highest level so won't want to define it in every API. So my question is how would I get the current user. i also have added my dependency.py file.

### Operating System

Windows

### FastAPI Version

0.109.2

### Pydantic Version

2.6.1

### Python Version

3.10.5

**Comment by jgould22** [comment, marked as answer]
The result of router based dependencies is discarded after the dependency is run as it says in the docs [here](https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-in-path-operation-decorators/#dependencies-in-path-operation-decorators).

FastAPI is designed around dependency injection for the path operations so the idea is to give every request that needs a user the user dependency.

**Reply by henrymcl** [reply]
> The result of router based dependencies is discarded after the dependency is run as it says in the docs [here](https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-in-path-operation-decorators/#dependencies-in-path-operation-decorators).
> 
> FastAPI is designed around dependency injection for the path operations so the idea is to give every request that needs a user the user dependency.

If I have the same dependency in router as well as in the path operation (for getting current user), will the dependency be called twice or will it be cached?

**Reply by jymchng** [reply]
It will be cached. `Depends` function has a `use_cache` parameter which is by default `True`.

**Reply by henrymcl** [reply]
Sorry I have another question on the caching: it works very fine for simple function-based dependencies, but what about parameterized class dependencies? When `/admin` is visited RoleChecker is run for 3 times

```py
from typing import Annotated

import uvicorn
from fastapi import Depends, FastAPI, HTTPException, status

app = FastAPI()


class RoleChecker:
    def __init__(self, allowed_role: str):
        self.allowed_role = allowed_role

    def __call__(self, role: str = "guest"):
        print(f"called with {role} <=> {self.allowed_role}")
        if role not in self.allowed_role:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail=f"User with role '{role}' is not allowed to access this endpoint.",
            )
        return role


# --- Using the Parameterized Dependencies in Routes ---


@app.get("/admin", dependencies=[Depends(RoleChecker("admin"))])
async def admin_dashboard(
    current_user_role: Annotated[str, Depends(RoleChecker("admin"))],
    current_user_role2: Annotated[str, Depends(RoleChecker("admin"))],
):
    return {"message": "Welcome to the admin dashboard!", "role": current_user_role}


@app.get("/")
async def index():
    """
    A public endpoint that doesn't require any specific role.
    """
    return {
        "message": "Welcome to the API. Try accessing /admin or /profile with a role parameter."
    }


if __name__ == "__main__":
    uvicorn.run("poc:app", host="0.0.0.0", port=8001, reload=True)
```

**Reply by jymchng** [reply]
Hi

It must be ran at least twice for your use case.

The dependency on the `@app.get` decorator is meant for the route even before the dependencies for the endpoint parameter is injected.

And for the two parameters, e.g. `current_user_role` and the other one, I suppose you would want to check the role for each 'User' which is then injected into the endpoint function as different 'User' arguments.

From this context, the route dependency is a little redundant hence at the start I mentioned it as 'at least twice'.
